### PR TITLE
[EXPLORER] Managing hiding behavior for tray notification icons. [WIP]

### PR DESCRIPTION
**Pull request by Getequ**
Here UI part only, selected behaviors not saved.
Also fixed saving changes for parameters "Show seconds" and "Group buttons"

### CODE REVIEW

File: sdk/include/psdk/windowsx.h [hunk 1]
```diff
@@ -509,7 +509,7 @@
 #define Static_SetIcon(hwndCtl,hIcon) ((HICON)(UINT)(DWORD)SendMessage((hwndCtl),STM_SETICON,(WPARAM)(HICON)(hIcon),0))
 #define Static_SetText(hwndCtl,lpsz) SetWindowText((hwndCtl),(lpsz))
 #define SubclassDialog(hwndDlg,lpfn) ((DLGPROC)SetWindowLongPtr(hwndDlg,DWLP_DLGPROC,(LPARAM)(DLGPROC)(lpfn)))
-#define SubclassWindow(hwnd,lpfn) ((WNDPROC)SetWindowLongPtr((hwnd),GWLP_WNDPROC,(LPARAM)(WNDPROC)(lpfn)))
+//#define SubclassWindow(hwnd,lpfn) ((WNDPROC)SetWindowLongPtr((hwnd),GWLP_WNDPROC,(LPARAM)(WNDPROC)(lpfn)))
```
Comment (Getequ) [inline sdk/include/psdk/windowsx.h]: this macro has conflict and used nowhere. May be delete it?

Comment (learn-more) [reply sdk/include/psdk/windowsx.h]: No, it should not be removed.

Comment (HBelusca) [reply sdk/include/psdk/windowsx.h]: Has conflicts? With which other definitions?

File: base/shell/explorer/notifyiconscust.cpp [hunk 2]
```diff
@@ -20,31 +21,248 @@
...
+    lvItem.cchTextMax = 100;
+    lvItem.pszText = state;
+
+    LVITEM lvItem2;
+    ZeroMemory(&lvItem2, sizeof(lvItem));
+    lvItem2.mask = LVIF_PARAM;
+    lvItem2.iSubItem = 0;
+    lvItem2.lParam = 2;
+
+    for(int i = 0; i< count; i++)
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: "for(space)(etc...)"

File: base/shell/explorer/notifyiconscust.cpp [hunk 3]
```diff
@@ -20,31 +21,248 @@
...
+    {
+        lvItem2.iItem = i;
+        ListView_SetItem(hListView, &lvItem2);
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: Why using two different lvItems?

File: base/shell/explorer/notifyiconscust.cpp [hunk 4]
```diff
@@ -20,31 +21,248 @@
...
     switch(Message)
     {
-        case WM_INITDIALOG:
-        
-        return TRUE;
+        case WM_INITDIALOG: 
+            InitializeListView(hwnd);
+            break;
+        case WM_NOTIFY:
+            {
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: There is a level of indentation too much.

File: base/shell/explorer/notifyiconscust.cpp [hunk 5]
```diff
@@ -20,31 +21,248 @@
...
+            switch(LOWORD(wParam))
+            {
+                case IDOK:
+                    EndDialog(hwnd, IDOK);
+                    break;
+                case IDCANCEL:
+                    EndDialog(hwnd, IDCANCEL);
+                    break;
+                case IDC_NOTIFICATION_BEHAVIOUR:
+                    {
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: No need to add this extra level of indentation.

File: base/shell/explorer/notifyiconscust.cpp [hunk 6]
```diff
@@ -20,31 +21,248 @@
...
     }
     return TRUE;
 }
 
-VOID ShowCustomizeNotifyIcons(HINSTANCE hInst, HWND hExplorer)
+VOID SetNotifyIcons(HWND hDialog, HWND hTrayNotify)
+{
+    HWND hListView = GetDlgItem(hDialog, IDC_NOTIFICATION_LIST);
+    HWND hSysPager = FindWindowEx(hTrayNotify, NULL, L"SysPager", NULL);
+    HWND hToolbar  = FindWindowEx(hSysPager, NULL, L"ToolbarWindow32", NULL);
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: Isn't there a way to retrieve the correct sub-window directly from the classes? Cc @gigaherz .

File: base/shell/explorer/notifyiconscust.cpp [hunk 7]
```diff
@@ -20,31 +21,248 @@
...
+    
+    HIMAGELIST tbImageList = (HIMAGELIST )SendMessage(hToolbar, TB_GETIMAGELIST, 0, 0);
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: Useless space after "HIMAGELIST" in the cast.

File: base/shell/explorer/notifyiconscust.cpp (hunk 6 above)
Comment (Getequ) [reply base/shell/explorer/notifyiconscust.cpp]: this is a standard win32 window classes, not a custom classes with unique name. It's lucky that hTrayNotify has only one sysPager and pager has only one toolbar :)

File: sdk/include/psdk/windowsx.h (hunk 1 above)
Comment (Getequ) [reply sdk/include/psdk/windowsx.h]: Yes, building failed with message "macro SubclassWindow requires 2 arguments, but only 1 given".
SubclassWindow function have one argument.

**Comment by Getequ** [comment]
@HBelusca  all done!

File: base/shell/explorer/notifyiconscust.cpp (hunk 3 above)
Comment (Getequ) [reply base/shell/explorer/notifyiconscust.cpp]: no need to whole reinitialize them on every iteration

File: base/shell/explorer/notifyiconscust.cpp (hunk 6 above)
Comment (HBelusca) [reply base/shell/explorer/notifyiconscust.cpp]: I was asking this because these HWNDs are just members of the corresponding CSysPager... classes, so I was wondering whether you could somehow do a pSysPager->m_hWnd somehow (with the correct names to be determined...) and/or use a correct getter function for this...

File: sdk/include/psdk/windowsx.h (hunk 1 above)
Comment (HBelusca) [reply sdk/include/psdk/windowsx.h]: > SubclassWindow function have one argument.

Because you are seeing the override from atlwin.h .
See the last results in: https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=SubclassWindow
The correct way to proceed would be to include the windowsx.h before atlwin.h & ATL consorts.

Comment (ThFabba) [reply sdk/include/psdk/windowsx.h]: I don't think that makes a difference, ATL does a push_macro/pop_macro on it. Just `#undef SubclassWindow` after including windowsx.h.

File: base/shell/explorer/traywnd.cpp [hunk 8]
```diff
@@ -2778,6 +2778,10 @@ class CTrayWindow :
             SetWindowPos(hWndInsertAfter, 0, 0, 0, 0, SWP_NOACTIVATE | SWP_NOMOVE | SWP_NOSIZE | SWP_SHOWWINDOW);
         }
 
+        g_TaskbarSettings.bHideInactiveIcons = newSettings->bHideInactiveIcons;
+        g_TaskbarSettings.bShowSeconds = newSettings->bShowSeconds;
+        g_TaskbarSettings.bGroupButtons = newSettings->bGroupButtons;
```
Comment (ThFabba) [inline base/shell/explorer/traywnd.cpp]: These last two lines don't seem related to tray icon hiding. What's their purpose?

File: base/shell/explorer/notifyiconscust.cpp [hunk 9]
```diff
@@ -20,31 +21,246 @@
...
+    lviBehavior.mask = LVIF_TEXT;
+    lviBehavior.iSubItem = 1;
+    lviBehavior.cchTextMax = 100;
+    lviBehavior.pszText = szBehavior;
+
+    LVITEM lviName;
+    ZeroMemory(&lviName, sizeof(lviName));
+    lviName.mask = LVIF_PARAM;
+    lviName.iSubItem = 0;
+    lviName.lParam = 2;
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: It should actually be OK to simplify these two things using a single LVITEM structure.
You zero it out as you do, then you set the members that do not change (lParam, cchTextMax and pszText), then, in the for loop and before you call each ListView_SetItem, you set (for the first one), iItem to i, iSubItem to 0 and mask to LVIF_PARAM, then (for the second one) you keep the iItem, you change iSubItem to 1 and mask to LVIF_TEXT.

File: base/shell/explorer/traywnd.cpp (hunk 8 above)
Comment (Getequ) [reply base/shell/explorer/traywnd.cpp]: just to save this flag too. yes, they not related to tray icons

File: base/shell/explorer/notifyiconscust.cpp [hunk 10]
```diff
@@ -20,31 +21,246 @@
...
+    HWND hListView = GetDlgItem(hwnd, IDC_NOTIFICATION_LIST);
+    int count = ListView_GetItemCount(hListView);
+    WCHAR szBehavior[100];
+    LoadStringW(NULL, IDS_NOTIF_BEH_HIDE_INACTIVE, szBehavior, _countof(szBehavior));
+
+    LVITEM lviBehavior;
+    ZeroMemory(&lviBehavior, sizeof(lviBehavior));
+    lviBehavior.mask = LVIF_TEXT;
+    lviBehavior.iSubItem = 1;
+    lviBehavior.cchTextMax = 100;
```
[the remark this answers is not in the document]
Comment (learn-more) [reply base/shell/explorer/notifyiconscust.cpp]: `CStringW` please

File: sdk/include/psdk/windowsx.h (hunk 1 above)
Comment (Getequ) [reply sdk/include/psdk/windowsx.h]: Ok, i add #undef SubclassWindow

File: base/shell/explorer/notifyiconscust.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Getequ) [reply base/shell/explorer/notifyiconscust.cpp]: i actually load it using LoadStringW with _countof .  100 was a copy-paste waste )

**Comment by ThFabba** [comment]
So... I'm a bit allergic to implementing UI with no functionality behind it. We already have lots of buttons and dialogs that do nothing, and they're all really confusing for users.

What can we do to avoid that? Actually implementing the feature would be one way. Somehow clearly indicating to the user that this is unimplemented could also work. Suggestions welcome.

File: base/shell/explorer/notifyiconscust.cpp (hunk 6 above)
Comment (Getequ) [reply base/shell/explorer/notifyiconscust.cpp]: Yeah, not a perfect way.  Tried to get toolbar HWND from CSysPager with functions like CTrayNotifyWnd_GetToolbar(m_TrayNotifyInstance) -> CSysPagerWnd_GetToolbar(&m_pager) -> Toolbar.m_hWnd field give me not null handle but it has very weird behavior:  TB_BUTTONCOUNT and Toolbar.GetButtonCount() return zero, but CNotifyToolbar.GetVisibleButtonCount() return nonzero value.

**Comment by Getequ** [comment]
@ThFabba  yes, there is a reason in your words. But i see that even not big PR can hang for months sinking in discussions, even if they have Approved mark.

File: base/shell/explorer/notifyiconscust.cpp (hunk 6 above)
Comment (yagoulas) [reply base/shell/explorer/notifyiconscust.cpp]: This is definitely an ugly way to implement an obscure feature...

File: base/shell/explorer/traywnd.cpp (hunk 8 above)
Comment (Getequ) [reply base/shell/explorer/traywnd.cpp]: oops, sorry ) rollback this

**Comment by HBelusca** [comment]
You can either add the new UI elements ni the resources but hide them (using a `| NOT WS_VISIBLE` for each of them in the resources), and/or add button handling to pop up a "not yet implemented" message box.

**Comment by Getequ** [comment]
No, this is also not an option) i better finish it with as possible functionality as i can. I recently found this issue https://jira.reactos.org/browse/CORE-10849 - this was VERY helpful

**Comment by learn-more** [comment]
So where are we going with this?

**Comment by Getequ** [comment]
I stuck with saving binary registy values. And also revealed new bug - importing files with big HEX registry values wrongly counting and read lines (regproc.c / processRegLinesW).

**Comment by gonzoMD** [comment]
Maybe @katahiromz can help you with that. He made some great explorer/shell fixes recently.

File: base/shell/explorer/resource.h [hunk 11]
```diff
@@ -102,6 +102,15 @@
...
 #define IDS_TASKBAR_STARTMENU_PROP_CAPTION 810
 
+#define IDS_NOTIF_BEH_NAME                 1000
+#define IDS_NOTIF_BEH_BEHAVIOR             1001
+#define IDS_NOTIF_BEH_HIDE_INACTIVE        1002
+#define IDS_NOTIF_BEH_HIDE                 1003
+#define IDS_NOTIF_BEH_SHOW                 1004
+
+#define IDS_NOTIF_BEH_CURRENT              1403
+#define IDS_NOTIF_BEH_PAST                 1404
```
Comment (HBelusca) [inline base/shell/explorer/resource.h]: 👍  for Brout-Englert-Higgs jkjk 😄

File: base/shell/explorer/notifyiconscust.cpp [hunk 12]
```diff
@@ -20,31 +21,244 @@
...
+    HWND hCombo = GetDlgItem(hwnd, IDC_NOTIFICATION_BEHAVIOUR);
+
+    if (ComboBox_GetCount(hCombo) == 0)
+    {
+        WCHAR sShow[100];
+        WCHAR sHide[100];
+        WCHAR sInactive[100];
+        LoadStringW(NULL, IDS_NOTIF_BEH_SHOW, sShow, _countof(sShow));
+        LoadStringW(NULL, IDS_NOTIF_BEH_HIDE, sHide, _countof(sHide));
+        LoadStringW(NULL, IDS_NOTIF_BEH_HIDE_INACTIVE, sInactive, _countof(sInactive));
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: I'm hesitating in using CStringW class for the resource strings here (they would automatically deal correctly with the string sizes). @learn-more , any advice on that?

File: base/shell/explorer/notifyiconscust.cpp [hunk 13]
```diff
@@ -20,31 +21,244 @@
...
+
+VOID SetItemBehavior(HWND hwnd, HWND hCombo)
+{
+    int iState = ComboBox_GetCurSel(hCombo);
+    int iItem = (int)ComboBox_GetItemData(hCombo, 0);
+    HWND hListView = GetDlgItem(hwnd, IDC_NOTIFICATION_LIST);
+
+    WCHAR szBehavior[100];
+    int resId = iState == 0 ? IDS_NOTIF_BEH_SHOW : (iState == 1 ? IDS_NOTIF_BEH_HIDE : IDS_NOTIF_BEH_HIDE_INACTIVE);
+    LoadStringW(NULL, resId, szBehavior, _countof(szBehavior));
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: Perhaps some CStringW usage? (see comment above)

File: base/shell/explorer/notifyiconscust.cpp [hunk 14]
```diff
@@ -20,31 +21,244 @@
...
+
+    ShowWindow(hCombo, SW_HIDE);
+}
+
+VOID ResetBehaviors(HWND hwnd)
+{
+    HWND hListView = GetDlgItem(hwnd, IDC_NOTIFICATION_LIST);
+    int count = ListView_GetItemCount(hListView);
+    WCHAR szBehavior[100];
+    LoadStringW(NULL, IDS_NOTIF_BEH_HIDE_INACTIVE, szBehavior, _countof(szBehavior));
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: Ditto

File: base/shell/explorer/notifyiconscust.cpp [hunk 15]
```diff
@@ -20,31 +21,244 @@
...
-        return TRUE;
+        case WM_INITDIALOG: 
+            InitializeListView(hwnd);
+            break;
+        case WM_NOTIFY:
+            if (LOWORD(wParam) == IDC_NOTIFICATION_LIST && ((LPNMHDR)lParam)->code == NM_CLICK) 
+            {
+                ShowBehaviorCombo(hwnd, (LPNMITEMACTIVATE)lParam);
+            }
+            else return FALSE;
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: Because first if-block was inside brackets, the return-FALSE should be inside brackets too (cf. our coding style).

File: base/shell/explorer/notifyiconscust.cpp [hunk 16]
```diff
@@ -20,31 +21,244 @@
...
 {
     switch(Message)
     {
-        case WM_INITDIALOG:
-        
-        return TRUE;
+        case WM_INITDIALOG: 
+            InitializeListView(hwnd);
+            break;
+        case WM_NOTIFY:
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: Please add one new-line between the different 'cases' to make the code more easily readable.

File: base/shell/explorer/notifyiconscust.cpp [hunk 17]
```diff
@@ -20,31 +21,244 @@
...
+    HIMAGELIST lvImageList = ImageList_Duplicate(tbImageList);
+    ListView_SetImageList(hListView, lvImageList, LVSIL_SMALL);
+
+    WCHAR sShow[100];
+    WCHAR sHide[100];
+    WCHAR sInactive[100];
+
+    LoadStringW(NULL, IDS_NOTIF_BEH_SHOW, sShow, _countof(sShow));
+    LoadStringW(NULL, IDS_NOTIF_BEH_HIDE, sHide, _countof(sHide));
+    LoadStringW(NULL, IDS_NOTIF_BEH_HIDE_INACTIVE, sInactive, _countof(sInactive));
```
Comment (HBelusca) [inline base/shell/explorer/notifyiconscust.cpp]: CStringW?

File: base/shell/explorer/syspager.cpp [hunk 18]
```diff
@@ -21,12 +21,6 @@
 
 #include "precomp.h"
 
-struct InternalIconData : NOTIFYICONDATA
-{
-    // Must keep a separate copy since the original is unioned with uTimeout.
-    UINT uVersionCopy;
-};
-
```
Comment (HBelusca) [inline base/shell/explorer/syspager.cpp]: Why isn't this needed anymore?

File: base/shell/explorer/notifyiconscust.cpp (hunk 12 above)
Comment (learn-more) [reply base/shell/explorer/notifyiconscust.cpp]: Why would you hestitate?
If you (for some obscure reason) don't want to use CStrings, you can always use the LoadStringW technique of passing it a pointer, and itll point that to the raw resource string (but be careful, it is not always null terminated!)

File: base/shell/explorer/resource.h (hunk 11 above)
Comment (Getequ) [reply base/shell/explorer/resource.h]: what?

File: base/shell/explorer/syspager.cpp (hunk 18 above)
Comment (Getequ) [reply base/shell/explorer/syspager.cpp]: No, it is moved to shell/explorer/precomp.h

File: base/shell/explorer/notifyiconscust.cpp (hunk 15 above)
Comment (Getequ) [reply base/shell/explorer/notifyiconscust.cpp]: ok

File: base/shell/explorer/notifyiconscust.cpp (hunk 16 above)
Comment (Getequ) [reply base/shell/explorer/notifyiconscust.cpp]: ok

**Comment by Getequ** [comment]
@gonzoMD hi, for now i waiting for accepting patch in Wine for reading imageList ( fix in #619 ). Without them this window will not look fine on ROS, only on Windows - i tested it by copying compiled explorer to XP in VBox

**Comment by katahiromz** [comment]
Usage of CStringW:
```c++
CStringW str(MAKEINTRESOURCEW(IDS_SOMETHING));
MessageBoxW(NULL, (LPCWSTR)str, NULL, 0);
```
Or:
```c++
CStringW str;
str.LoadString(IDS_SOMETHING);
MessageBoxW(NULL, (LPCWSTR)str, NULL, 0);
```
Or:
```c++
CStringW str;
str.Format(IDS_SOMETHING, 3.14)
MessageBoxW(NULL, (LPCWSTR)str, NULL, 0);
```

**Comment by katahiromz** [comment]
For more details, see "sdk/lib/atl/cstringt.h".

File: base/shell/explorer/notifyiconscust.cpp (hunk 12 above)
Comment (katahiromz) [reply base/shell/explorer/notifyiconscust.cpp]: Usage:
```cpp
CStringW str...(MAKEINTRESOURCEW(IDS_...));
```

**Comment by Saibamen** [comment]
Please mark this Pull Request as Draft if is still WIP: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request
